### DISCUSSION TITLE
Custom build.rs cannot find a dependency

### DISCUSSION THREAD

**Discussion by 4JX** [ipetkov/crane#391] [Q&A] [answered]
Having some trouble with adding gtk related dependencies to my repo ([flake here]( https://github.com/4JX/L5P-Keyboard-RGB/blob/19.2/flake.nix))

Running `nix run . -L --` gives
```
error: builder for '/nix/store/vncyapz464vqiz2s362q0m25k15l62r5-legion-kb-rgb-0.19.0.drv' failed with exit code 101;
       last 10 log lines:
       > ++ command cargo build --release --message-format json-render-diagnostics --locked
       >    Compiling legion-kb-rgb v0.19.0 (/build/source/app)
       >    Compiling legion-rgb-driver v0.1.0 (/build/source/driver)
       > error: failed to run custom build command for `legion-kb-rgb v0.19.0 (/build/source/app)`
       > note: To improve backtraces for build dependencies, set the CARGO_PROFILE_RELEASE_BUILD_OVERRIDE_DEBUG=true environment variable to enable debug information generation.
       >
       > Caused by:
       >   process didn't exit successfully: `/build/source/target/release/build/legion-kb-rgb-571b2982fa2eefd1/build-script-build` (exit status: 127)
       >   --- stderr
       >   /build/source/target/release/build/legion-kb-rgb-571b2982fa2eefd1/build-script-build: error while loading shared libraries: libgobject-2.0.so.0: cannot open shared object file: No such file or directory
       For full logs, run 'nix log /nix/store/vncyapz464vqiz2s362q0m25k15l62r5-legion-kb-rgb-0.19.0.drv'.
error: 1 dependencies of derivation '/nix/store/5ffy9w745qkhydz6bg3c2yscqb0hl0l7-legion-kb-rgb.drv' failed to build
```

Tried adding `gobject-introspection` to `buildInputs`  (what I assumed would work) and then `glib` for redundancy (which seems to contain the lib if queried with `nix-index -c nix-locate --top-level libgobject-2.0`) but neither worked. Doesn't feel like this is specifically an issue with `crane` itself, but I'm really not sure why its not picking up on the lib's existence.

**Comment by ipetkov** [comment]
You may be missing a `nativeBuildInputs = [ pkgs.pkg-config ];`? `pkg-config` is the standard way of having different build scripts find out about where particular libraries are (which is especially important for building with Nix since those libraries won't be at `/usr/lib` or equivalent

**Comment by 4JX** [comment]
Wouldn't seem like it as I already [do have that](https://github.com/4JX/L5P-Keyboard-RGB/blob/ce2639fc67942a77fd39c3ce427542785b9568e7/flake.nix#L148).

**Reply by ipetkov** [reply]
Hmm I'm not entirely sure then. I would investigate how the build script (or rather the crate it delegates to) is looking for libraries and how to nudge that in the right direction

**Comment by dpc** [comment]
Are you maybe using mold with a wrapper?

**Reply by 4JX** [reply]
`mold` is used through adding a few env vars
```
MOLD_PATH = "${pkgs.mold.out}/bin/mold";
RUSTFLAGS = "-Clink-arg=-fuse-ld=${MOLD_PATH} -Clinker=clang";
LIBCLANG_PATH = "${pkgs.llvmPackages.libclang.lib}/lib";
```

And [`pkgs.mold`](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/development/tools/mold/default.nix) just pulls and builds from source.

Or are you referring to using `mold` in any form?

**Reply by dpc** [reply]
https://github.com/fedimint/fedimint/pull/3228/files

**Reply by 4JX** [reply, marked as answer]
Good lead, landed upon fixing it by removing the manual `mold` setup vars and instead using `stdenv = pkgs.stdenvAdapters.useMoldLinker pkgs.stdenv` in the dependency and main package derivations.

**Reply by dpc** [reply]
BTW. There's a PR https://github.com/NixOS/nixpkgs/pull/257158 to nixpkgs to make using wrapped mold easier, which should benefit Rust users in particular.
